src/analysis/normed_space/mazur_ulam.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2020 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Yury Kudryashov
+-/
+import analysis.normed_space.point_reflection
+import topology.instances.real_vector_space
+
+/-!
+# Mazur-Ulam Theorem
+
+Mazur-Ulam theorem states that an isometric bijection between two normed spaces over `ℝ` is affine.
+Since `mathlib` has no notion of an affine map (yet?), we formalize it in two definitions:
+
+* `isometric.to_real_linear_equiv_of_map_zero` : given `E ≃ᵢ F` sending `0` to `0`,
+  returns `E ≃L[ℝ] F` with the same `to_fun` and `inv_fun`;
+* `isometric.to_real_linear_equiv` : given `f : E ≃ᵢ F`,
+  returns `g : E ≃L[ℝ] F` with `g x = f x - f 0`.
+
+The formalization is based on [Jussi Väisälä, *A Proof of the Mazur-Ulam Theorem*][Vaisala_2003].
+
+## Tags
+
+isometry, affine map, linear map
+-/
+
+variables {E : Type*} [normed_group E] [normed_space ℝ E]
+  {F : Type*} [normed_group F] [normed_space ℝ F]
+
+open set
+
+noncomputable theory
+
+namespace isometric
+
+/-- If an isometric self-homeomorphism of a normed vector space over `ℝ` fixes `x` and `y`,
+then it fixes the midpoint of `[x, y]`. This is a lemma for a more general Mazur-Ulam theorem,
+see below. -/
+lemma midpoint_fixed {x y : E} :
+  ∀ e : E ≃ᵢ E, e x = x → e y = y → e (midpoint ℝ x y) = midpoint ℝ x y :=
+begin
+  set z := midpoint ℝ x y,
+  set s := { e : E ≃ᵢ E | e x = x ∧ e y = y },
+  set c := ⨆ e : s, dist (e z) z,
+  haveI : nonempty s := ⟨⟨isometric.refl E, rfl, rfl⟩⟩,
+  have A : bdd_above (range $ λ e : s, dist (e z) z),
+  { refine ⟨dist x z + dist x z, forall_range_iff.2 $ λ e, _⟩,
+    calc dist (e z) z ≤ dist (e z) x + dist x z : dist_triangle (e z) x z
+    ... = dist (e x) (e z) + dist x z : by erw [e.2.1, dist_comm]
+    ... = _ : by erw [e.1.dist_eq x z] },
+  suffices : c ≤ 0,
+  { intros e hx hy,
+    have A : dist (e z) z ≤ c, from @le_csupr _ _ _ _ A ⟨e, hx, hy⟩,
+    exact dist_le_zero.1 (le_trans A this) },
+  suffices : c ≤ c / 2, { linarith },
+  apply csupr_le,
+  set g : E ≃ᵢ E := point_reflection z,
+  rintros ⟨e, he⟩,
+  apply le_div_of_mul_le (zero_lt_two : 0 < (2:ℝ)),
+  have : ((e.trans g).trans e.symm).trans g ∈ s,
+    by split; simp [he.1, he.2, (e.symm_apply_eq).2 he.1.symm, (e.symm_apply_eq).2 he.2.symm],
+  dsimp,
+  rw [mul_comm, dist_comm, ← point_reflection_dist_self_real, ← e.symm.dist_eq, e.symm_apply_apply,
+    ← point_reflection_dist_fixed],
+  exact @le_csupr _ _ _ _ A ⟨_, this⟩
+end
+
+/-- Mazur-Ulam Theorem: if `f` is an isometric bijection between two normed vector spaces
```

Comment:
Instead of giving the theorem as a definition, could you first prove a version saying `is_linear_map (f : E -> F)`, and then build the continuous linear equiv using this one? I know `is_add_group_hom`, `is_linear_map` and so on are semi-deprecated (and in particular their use as classes should be strictly forbidden) but in this kind of situation it would really clarify the statement.

Reply:
We have an API for `add_monoid_hom.to_real_linear_map`, not `add_monoid_hom.is_real_linear_map`. I added a comment, and I can add an `example` converting this `continuous_linear_equiv` to `is_linear_map`.